[SCSS | Sass] Fix hyphens in interpolated indents

bgriffith opened the pull request:
Fixes parse issue when parser comes across hyphens in interpolated indents i.e.

``` scss
.-#{$i}
.-#{$s1}.-#{$s2}
.-#{$s1}-#{$s2}
```

Also removes some unused variables (interpolation array, isInt bool)

re https://github.com/sasstools/sass-lint/issues/644

tonyganch commented on src/sass/parse.js:
```diff
@@ -1786,13 +1785,23 @@ function checkIdent(i) {
+
+  if (!wasIdent &&
+      tokens[start].type !== TokenType.Asterisk &&
+      tokens[start].type !== TokenType.HyphenMinus) return 0;
```
That will parse `.-, a {}` as valid code, where `.-` will be a class named `-`.
The issue is somewhat related to #114 and I can't think of a fast and correct solution to propose.

As a hackish fix you can modify `checkClass` and `checkShash` functions to check not only interpolations/idents but also hyphens before/between them. This will fix this specific case and not break anything else.
What do you think?

bgriffith replied:
I'll have another look at this one tomorrow :+1:

bgriffith commented:
@tonyganch I've updated this PR to modify the `checkClass` / `getClass` functions instead. Let me know if it's going in the right direction.

Also could potentially tidy it up with a simple `checkHyphen` / `getHyphen`. What you do you think?

bertusgroenewegen commented:
Please fix :+1:

bertusgroenewegen commented:
@tonyganch could you solve this issue ? Thanks alot!

bgriffith commented:
Hey @tonyganch - I've updated this with the latest on dev and fixed some JSCS issues. Happy to revisit if you have some feedback! Would really like to fix this one! It's a biggie! :+1:

tonyganch commented:
I hear you, guys. Please give me a day or two to finish some other work and I'll get to you.

tonyganch commented:
TIL `--\__\--` is a valid class name in Sass!

It's invalid in plain CSS though. Do you have any insight on whether that's done in Sass by design or accident?
Since Sass code is supposed to be compiled into (valid) CSS, I'd suggest this is error in preprocessor.

tonyganch commented:
@bgriffith, what's the plan when in a week or a month someone comes to you and says that `._#{$s1}__#{$s2}` or `._#{$s1}111#{$s2}` does not work for him?

bgriffith commented:
Hey @tonyganch - Thanks for the feedback!

Re - `--\__\--` - I would think this is a bug in Sass but that's only a guess - sorry! 😄 

Having just tested `._#{$s1}__#{$s2}`, pleased to say it works fine as `_` is treated as an ident. The `-` seems to be the only valid selector character that requires the extra work (may be wrong here).

The second example `._#{$s1}111#{$s2}` fails but not due to this PR as far as I can tell and does instead raise another parse issue to do with having numbers trailing an interpolation in a selector.

`.#{$s1}123` is a simplified way to see this. I think this should be solved in a separate PR.

PS I can add some more tests using `_` to this PR.

Thought?

tonyganch commented on src/sass/parse.js:
```diff
@@ -1115,15 +1115,30 @@ function getBrackets() {
+    }
+    else if (tokens[i].type === TokenType.HyphenMinus) i += 1;
+    else break;
+  }
```
According to CSS validator, this is a valid class name: `-a---a--`, but this is not: `--a`. It means that there can be any number of hyphens just not at the beginning. To make the code above disallow situation with multiple hyphens, I'd propose removing `wasIdent` variable and making something like that:

``` js
// Check for `-` at beginning.
if (tokens[i].type === TokenType.HyphenMinus) i+= 1;

// Original lines.
if (l = checkIdentOrInterpolation(i)) i += l;
else return 0;

// Here goes your loop, just without `wasIdent` thing.
while (...) {...}
```

bgriffith replied:
Looks good to me. Have you found a detailed spec for valid css selectors? I can't find anything in the docs!

tonyganch replied:
From https://www.w3.org/TR/selectors4/#class-html:

> The class selector is given as a full stop (. U+002E) immediately followed by an identifier.

From https://www.w3.org/TR/CSS21/syndata.html#value-def-identifier:

> In CSS, identifiers (including element names, classes, and IDs in selectors) can contain only the characters [a-zA-Z0-9] and ISO 10646 characters U+00A0 and higher, plus the hyphen (-) and the underscore (_); they cannot start with a digit, two hyphens, or a hyphen followed by a digit. Identifiers can also contain escaped characters and any ISO 10646 character as a numeric code (see next item). For instance, the identifier "B&W?" may be written as "B&W\?" or "B\26 W\3F".

tonyganch replied:
Should I put this into GPE docs? :)

bgriffith committed:
[scss] Fix hyphens in interpolated idents

bgriffith committed:
[sass] Fix hyphens in interpolated idents

bgriffith commented:
Added your suggestion from https://github.com/tonyganch/gonzales-pe/pull/179#discussion_r70787278 :+1:
